evm_arithmetization/src/generation/mod.rs

```diff
@@ -310,38 +312,28 @@ pub fn generate_traces<F: RichField + Extendable<D>, const D: usize>(
 }
 
 fn simulate_cpu<F: Field>(state: &mut GenerationState<F>) -> anyhow::Result<()> {
-    let halt_pc = KERNEL.global_labels["halt"];
+    state.run_cpu()?;
+
+    let pc = state.registers.program_counter;
+    // Padding
+    let mut row = CpuColumnsView::<F>::default();
+    row.clock = F::from_canonical_usize(state.traces.clock());
+    row.context = F::from_canonical_usize(state.registers.context);
+    row.program_counter = F::from_canonical_usize(pc);
+    row.is_kernel_mode = F::ONE;
+    row.gas = F::from_canonical_u64(state.registers.gas_used);
+    row.stack_len = F::from_canonical_usize(state.registers.stack_len);
 
     loop {
-        // If we've reached the kernel's halt routine, and our trace length is a power
-        // of 2, stop.
-        let pc = state.registers.program_counter;
-        let halt = state.registers.is_kernel && pc == halt_pc;
-        if halt {
-            log::info!("CPU halted after {} cycles", state.traces.clock());
-
-            // Padding
-            let mut row = CpuColumnsView::<F>::default();
-            row.clock = F::from_canonical_usize(state.traces.clock());
-            row.context = F::from_canonical_usize(state.registers.context);
-            row.program_counter = F::from_canonical_usize(pc);
-            row.is_kernel_mode = F::ONE;
-            row.gas = F::from_canonical_u64(state.registers.gas_used);
-            row.stack_len = F::from_canonical_usize(state.registers.stack_len);
-
-            loop {
-                state.traces.push_cpu(row);
-                row.clock += F::ONE;
-                if state.traces.clock().is_power_of_two() {
-                    break;
-                }
-            }
-
-            log::info!("CPU trace padded to {} cycles", state.traces.clock());
-
-            return Ok(());
+        // If our trace length is a power of 2, stop.
```

Comment:
minor, but I'd adapt the two small comments following from the change with the new `State` trait, the first one explaining we're setting some initial values for padding rows (as it's not anymore within the loop), and the second being moved to the conditional break below